File: lib_bgp_simulator/engine/as_classes/bgp/bgp_simple_as/propagate_funcs.py
from typing import List

from lib_caida_collector import AS

from ....ann_containers import LocalRIB
from ....ann_containers import RecvQueue
from .....enums import Relationships
from .....announcements import Announcement as Ann


def propagate_to_providers(self):
    """Propogates to providers"""

    send_rels = set([Relationships.ORIGIN, Relationships.CUSTOMERS])
    self._propagate(Relationships.PROVIDERS, send_rels)

def propagate_to_customers(self):
    """Propogates to customers"""

    send_rels = set([Relationships.ORIGIN,
                     Relationships.CUSTOMERS,
                     Relationships.PEERS,
                     Relationships.PROVIDERS])
    self._propagate(Relationships.CUSTOMERS, send_rels)

def propagate_to_peers(self):
    """Propogates to peers"""

    send_rels = set([Relationships.ORIGIN,
                     Relationships.CUSTOMERS])
    self._propagate(Relationships.PEERS, send_rels)

def _propagate(self,
               propagate_to: Relationships,
               send_rels: List[Relationships]):
    """Propogates announcements from local rib to other ASes

    send_rels is the relationships that are acceptable to send

    Later you can change this so it's not the local rib that's
    being sent. But this is just proof of concept.
    """

    for neighbor in getattr(self, propagate_to.name.lower()):
        for prefix, ann in self._local_rib.prefix_anns():
            if ann.recv_relationship in send_rels and not self._prev_sent(neighbor, ann):
                propagate_args = [neighbor, ann, propagate_to, send_rels]
                # Policy took care of it's own propagation for this ann
                if self._policy_propagate(*propagate_args):
                    continue
                else:
                    self._process_outgoing_ann(*propagate_args)

def _policy_propagate(*args, **kwargs) -> bool:
    """Custom policy propagation that can be overriden"""

    return False

def _prev_sent(*args, **kwargs) -> bool:
    """Don't resend anything for BGPAS. For this class it doesn't matter"""
    return False

def _process_outgoing_ann(self,
                          neighbor: AS,
                          ann: Ann,
                          propagate_to: Relationships,
                          send_rels: List[Relationships]):
    """Adds ann to the neighbors recv q"""

    # Add the new ann to the incoming anns for that prefix
    neighbor.receive_ann(ann)

File: bgpc/src/cpp_simulation_engine/policies/bgp_simple_policy/propagate_funcs.cpp
#include <iostream>
#include <algorithm>
#include <stdexcept>

#include "policy.hpp"
#include "bgp_simple_policy.hpp"
#include "as.hpp"

void BGPSimplePolicy::propagate_to_providers() {
    std::set<Relationships> send_rels = {Relationships::ORIGIN, Relationships::CUSTOMERS};
    propagate(Relationships::PROVIDERS, send_rels);
}

void BGPSimplePolicy::propagate_to_customers() {
    std::set<Relationships> send_rels = {Relationships::ORIGIN, Relationships::CUSTOMERS, Relationships::PEERS, Relationships::PROVIDERS};
    propagate(Relationships::CUSTOMERS, send_rels);
}

void BGPSimplePolicy::propagate_to_peers() {
    std::set<Relationships> send_rels = {Relationships::ORIGIN, Relationships::CUSTOMERS};
    propagate(Relationships::PEERS, send_rels);
}

void BGPSimplePolicy::propagate(Relationships propagate_to, const std::set<Relationships>& send_rels) {
    std::vector<std::weak_ptr<AS>> neighbors;

    auto as_shared = as.lock();
    if (!as_shared) {
        // Handle the case where the AS object is no longer valid
        throw std::runtime_error("Weak ref from policy to as no longer exists");
    }

    switch (propagate_to) {
        case Relationships::PROVIDERS:
            neighbors = as_shared->providers;
            break;
        case Relationships::PEERS:
            neighbors = as_shared->peers;
            break;
        case Relationships::CUSTOMERS:
            neighbors = as_shared->customers;
            break;
        default:
            throw std::runtime_error("Unsupported relationship type.");
    }

    for (const auto& neighbor_weak : neighbors) {
        for (const auto& [prefix, ann] : localRIB.prefix_anns()) {
            if (send_rels.find(ann->recv_relationship) != send_rels.end() && !prev_sent(neighbor_weak, ann)) {
                if (policy_propagate(neighbor_weak, ann, propagate_to, send_rels)) {
                    continue;
                } else {
                    process_outgoing_ann(neighbor_weak, ann, propagate_to, send_rels);
                }
            }
        }
    }
}

bool BGPSimplePolicy::policy_propagate(const std::weak_ptr<AS>& neighbor_weak, const std::shared_ptr<Announcement>& ann, Relationships propagate_to, const std::set<Relationships>& send_rels) {
    // This method simply returns false and does not use the neighbor_weak reference
    return false;
}

bool BGPSimplePolicy::prev_sent(const std::weak_ptr<AS>& neighbor_weak, const std::shared_ptr<Announcement>& ann) {
    // This method simply returns false and does not use the neighbor_weak reference
    return false;
}

void BGPSimplePolicy::process_outgoing_ann(const std::weak_ptr<AS>& neighbor_weak, const std::shared_ptr<Announcement>& ann, Relationships propagate_to, const std::set<Relationships>& send_rels) {
    auto neighbor = neighbor_weak.lock();
    if (!neighbor || !neighbor->policy) {
        throw std::runtime_error("weak ref no longer exists");
    }
    neighbor->policy->receive_ann(ann);
}


